codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsHttpPresignURLClientGenerator.java

```diff
@@ -121,19 +247,76 @@ public void writePresignClientType(GoWriter writer) {
+                    writer.openBlock("result, _, err := c.client.invokeOperation(ctx, $S, params, clientOptFns,", ")",
                             operationSymbol.getName(), () -> {
                                 writer.write("$L,", OperationGenerator
                                         .getAddOperationMiddlewareFuncName(operationSymbol));
                                 writer.write("c.$L,", CONVERT_TO_PRESIGN_MIDDLEWARE_NAME);
+
+                                // s3 should add a middleware where it switches to using unisgned payload if
+                                // input is a stream.
+                                if (isS3ServiceShape(model, serviceShape)) {
+                                    if (operationInputShape.members().stream().anyMatch(memberShape -> {
+                                        return memberShape.getMemberTrait(model, StreamingTrait.class).isPresent();
+                                    })) {
+                                        writer.write("$L,", addAsUnsignedPayloadName(operationSymbol.getName()));
+                                    }
+                                }
                             });
-                    writer.write("if err != nil { return ``, nil, err }");
+                    writer.write("if err != nil { return nil, err }");
                     writer.write("");
 
-                    writer.write("out := result.(*v4.PresignedHTTPRequest)");
-                    writer.write("return out.URL, out.SignedHeader, nil");
+                    writer.write("out := result.($P)", v4PresignedHTTPRequestSymbol);
+                    writer.write("return out, nil");
                 });
+        writer.write("");
+    }
+
+    private void writeS3AddAsUnsignedPayloadHelper(
+            GoWriter writer,
+            Model model,
+            SymbolProvider symbolProvider,
+            ServiceShape serviceShape,
+            OperationShape operationShape
+    ) {
+        // if service is not s3, return
+        if (!isS3ServiceShape(model, serviceShape)) {
+            return;
+        }
+
+        Symbol operationSymbol = symbolProvider.toSymbol(operationShape);
 
+        Shape operationInputShape = model.expectShape(operationShape.getInput().get());
+
+        // return if not streaming
+        if (operationInputShape.members().stream().noneMatch(memberShape -> {
+            return memberShape.getMemberTrait(model, StreamingTrait.class).isPresent();
+        })) { return; }
+
+        writer.openBlock("func $L(stack $P, options Options) error {", "}",
+                addAsUnsignedPayloadName(operationSymbol.getName()),
+                SymbolUtils.createPointableSymbolBuilder("Stack", SmithyGoDependency.SMITHY_MIDDLEWARE).build(),
+                () -> {
+                    writer.write("return $T(stack)", SymbolUtils.createValueSymbolBuilder(
+                            "AddUnsignedPayloadMiddleware", AwsGoDependency.AWS_SIGNER_V4).build());
```

Comment:
Does the compute sha middleware needed to be removed as well?  Or does the unsigned Payload middleware get added before that compute sha is run?

Reply:
looks like this was already fixed,